File: baking.java
/**********************
*
* In this problem, we are tasked with finding the area of a shape consisting
* only of sides parallel to the major axes, with the property that any vertical
* line drawn through the shape will divide it into at most 2 pieces. This property
* is vital to the solution presented below, which employs a scan-line approach to
* sum the area of each width-1 slice of the shape. Because any vertical line will
* divide the shape into at most 2 parts, we guarantee that a width-1 slice of the
* shape will be rectangular, given that corner coordinates are integers.
*
* Our strategy will be to maintain a scan-line, or sliding window, representing the
* vertical slice of the shape (low y and high y) at some x-coordinate, as we sweep it
* through the shape from left to right and adjust it accordingly. At each step, the 
* window will represent a width-1 slice of the shape whose area we can add to a running sum.
* We are guaranteed that the leftmost side of the shape has exactly 2 corners, and because
* of the property described above, we know that at any x coordinate there will be either
* 0, 2, or 4 corners. So, if we maintain a list of corners for each x coordinate, we can
* update our scan-line during the sweep! The logic behind this update and how to use the
* corner data is described in greater detail below.
*
**********************/
import java.util.*;

public class baking{
	//Offset value to add to each x-coordinate, translating the shape onto the positive side of the x-axis.
	//We do this so we can easily use the x-coordinate as an index into a 0-based array.
	static int OFFSET = 10000;
	
	public static void main(String[] args){
		Scanner reader = new Scanner(System.in);
		int times = reader.nextInt();
		for(int t = 1; t <= times; t++){
			int n = reader.nextInt();
			
			//Maintain these arrays for corner data
			int[] cnt = new int[20001]; //Number of corners with the same x-coordinate (at most 4)
			int[][] pts = new int[20001][4]; //pts[x] stores the y-coordinates of all corners that share that x coordinate (at most 4), sorted in increasing order
			
			//Find the x-interval in which the shape resides
			int start = 20001; //The leftmost x-coordinate
			int end = -1; //The rightmost x-coordinate
			
			for(int i = 0; i < n; i++){
				//Read the corner coordinate and translate it by the x-offset
				int x = reader.nextInt() + OFFSET;
				int y = reader.nextInt();
				
				//Update our start and end values (widen the x-interval if appropriate)
				start = Math.min(start, x);
				end = Math.max(end, x);
				
				//Add this y coordinate to the list of corners sharing this x coordinate
				pts[x][cnt[x]] = y;
				//Increment the number of corners here
				cnt[x]++;
				//Sort the y coordinates at this location
				Arrays.sort(pts[x], 0, cnt[x]);
			}
			
			//Initialize our scanline with the leftmost face of the shape
			int bot = pts[start][0]; //The low y-coordinate
			int top = pts[start][1]; //The high y-coordinate
			//We take the first slice for fre
			int sum = top-bot;
			
			//Iterate from start+1 to end, exclusive
			for(int i = start+1; i < end; i++){
				//Working corners
				int[] y = pts[i];
				
				//Update our scanline here
				if(cnt[i] == 2){ //2 corners share this coordinate
					//If this is the case, then only one side of our window will change.
					//Which side of the window it is depends on which side matches one of
					//the given corners. For example, if the low-y (bot) of our window is
					//equal to the lesser of the 2 corner y coordinates, then the bottom
					//of our window will elevate to the greater of the 2 corner y coordinates.
					if(y[0] == bot)
						bot = y[1];
					else if(y[1] == bot)
						bot = y[0];
					else if(y[0] == top)
						top = y[1];
					else
						top = y[0];
				}else if(cnt[i] == 4){ //4 corners share this coordinate
					//We can handle the low-y and high-y of our window seperately in this case.
					//Similar to the case where cnt[i] == 2, we flip-flop based on which value
					//in the corner y-coordinates is equal to this end of the window.
					
					//We check the low-y of our window against y[0] and y[1]
					if(y[0] == bot)
						bot = y[1];
					else
						bot = y[0];
					
					//We check the high-y of our window against y[2] and y[3]
					if(y[3] == top)
						top = y[2];
					else
						top = y[3];
				}
				
				//Add the slice to our total
				sum += top-bot;
			}
			
			//Output the result
			System.out.println("Batch #" + t + ": " + sum + " square centimeters");
		}
	}
}

File: baking.cpp
#include <stdio.h>
#include <stdlib.h>

using namespace std;

/*
This problem reduces to having n points that we want to create a polygon
from where every corner(vertex) of the polygon makes a 90 degree turn. If
we look at points on the same horizontal line they MUST pair up with the
next point on the same axis. We can use this same idea for the vertical axis.
Once we know the order of the points in our polygon we can easily compute
the area.
*/

struct point
{
    int x, y, vert, horz;
};
// Point array is global for sorting.
point pt[500];

// Compare points first by x value then y.
int compareXY(const void * a, const void * b)
{
    int xa = pt[*(int*)a].x;
    int xb = pt[*(int*)b].x;
    if(xa == xb)
    {
        int ya = pt[*(int*)a].y;
        int yb = pt[*(int*)b].y;
        return ya - yb;
    }
    return xa - xb;
}

// Compare points first by y value then x.
int compareYX(const void * a, const void * b)
{
    int ya = pt[*(int*)a].y;
    int yb = pt[*(int*)b].y;
    if(ya == yb)
    {
        int xa = pt[*(int*)a].x;
        int xb = pt[*(int*)b].x;
        return xa - xb;
    }
    return ya - yb;
}

int main()
{
    int tests;
    scanf("%d", &tests);

    for(int t = 1; t<=tests; t++)
    {
        int n;
        scanf("%d", &n);
        for(int i = 0; i<n; i++)
            scanf("%d %d", &pt[i].x, &pt[i].y);

        // We'll use this array to find the relative ordering of the points.
        int sorted[n];
        for(int i = 0; i<n; i++)
            sorted[i] = i;

        // Sort by x then y. Then pair every two points together.
        qsort(sorted, n, sizeof(int), compareXY);
        for(int i = 0; i<n; i+=2)
        {
            pt[sorted[i]].vert = sorted[i+1];
            pt[sorted[i+1]].vert = sorted[i];
        }

        // Do the same for the YX sorting.
        qsort(sorted, n, sizeof(int), compareYX);
        for(int i = 0; i<n; i+=2)
        {
            pt[sorted[i]].horz = sorted[i+1];
            pt[sorted[i+1]].horz = sorted[i];
        }

        // Now we have an ordering for our polygon.
        int area = 0;
        int prev = 0;
        int cur = pt[prev].horz;
        bool nextVert = true;
        while(cur != 0)
        {
            area += (pt[cur].x-pt[prev].x)*pt[cur].y;

            // Find the next vertex on the polygon.
            prev = cur;
            cur = nextVert ? pt[cur].vert : pt[cur].horz;

            // Alternate vertical and horizontal for 90 degree turns.
            nextVert = !nextVert;
        }
        // Our ordering might have been counterclock-wise so our area will be inverted.
        if(area < 0)
            area = -area;

        printf("Batch #%d: %d square centimeters\n", t, area);
    }
    return 0;
}
